AppCenter/AppCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -264,100 +273,163 @@ + (void)enableAutoVacuumInOpenedDatabase:(void *)db {
 }
 
-+ (NSArray<NSArray *> *)executeSelectionQuery:(NSString *)query inOpenedDatabase:(void *)db result:(int *)result {
++ (NSArray<NSArray *> *)executeSelectionQuery:(NSString *)query
+                             inOpenedDatabase:(void *)db
+                                       result:(int *)result
+                                   withValues:(nullable NSArray *)values {
   NSMutableArray<NSMutableArray *> *entries = [NSMutableArray<NSMutableArray *> new];
-  sqlite3_stmt *statement = NULL;
-  int prepareResult = sqlite3_prepare_v2(db, [query UTF8String], -1, &statement, NULL);
-  if (result != nil) {
-    *result = prepareResult;
-  }
-  if (prepareResult == SQLITE_OK) {
-
-    // Loop on rows.
-    while (sqlite3_step(statement) == SQLITE_ROW) {
-      NSMutableArray *entry = [NSMutableArray new];
-
-      // Loop on columns.
-      for (int i = 0; i < sqlite3_column_count(statement); i++) {
-        id value = nil;
-
-        /*
-         * Convert values.
-         * TODO: Add here any other type it needs.
-         */
-        switch (sqlite3_column_type(statement, i)) {
-        case SQLITE_INTEGER:
-          value = @(sqlite3_column_int(statement, i));
-          break;
-        case SQLITE_TEXT:
-          value = [NSString stringWithUTF8String:(const char *)sqlite3_column_text(statement, i)];
-          break;
-        default:
-          value = [NSNull null];
-          break;
-        }
-        [entry addObject:value];
-      }
-      if (entry.count > 0) {
-        [entries addObject:entry];
-      }
-    }
-    sqlite3_finalize(statement);
-  } else {
-    MSLogError([MSAppCenter logTag], @"Query \"%@\" failed with error: %d - %@", query, prepareResult,
-               [NSString stringWithUTF8String:sqlite3_errmsg(db)]);
+  int queryResult = [MSDBStorage executeQuery:query
+                             inOpenedDatabase:db
+                                   withValues:values
+                                   usingBlock:^(void *statement) {
+                                     int stepResult;
+
+                                     // Loop on rows.
+                                     while ((stepResult = sqlite3_step(statement)) == SQLITE_ROW) {
+                                       NSMutableArray *entry = [NSMutableArray new];
+
+                                       // Loop on columns.
+                                       for (int i = 0; i < sqlite3_column_count(statement); i++) {
+                                         NSObject *value = [MSDBStorage columnValueFromStatement:statement withIndex:i];
+                                         [entry addObject:value];
+                                       }
+                                       if (entry.count > 0) {
+                                         [entries addObject:entry];
+                                       }
+                                     }
+                                     if (stepResult != SQLITE_DONE) {
```

Comment:
Do we still return `SQLITE_OK` for the error? Maybe we can set an error code in `result`?

Reply:
That's a block. But return is missing yeah